common/src/main/java/net/opentsdb/threadpools/UserAwareThreadPoolExecutor.java

```diff
@@ -0,0 +1,352 @@
+      tsdb.getConfig().register(getConfigKey(CORE_THREAD_POOL_SIZE),
+          Runtime.getRuntime().availableProcessors() * 2, false,
+          "Core thread pool size. " + "Can be null to use the default.");
+    }
+
+    if (!tsdb.getConfig().hasProperty(getConfigKey(MAX_THREAD_POOL_SIZE))) {
+      tsdb.getConfig().register(getConfigKey(MAX_THREAD_POOL_SIZE),
+          Runtime.getRuntime().availableProcessors() * 2, false,
+          "Max thread pool size. " + "Can be null to use the default.");
+    }
+
+    if (!tsdb.getConfig().hasProperty(getConfigKey(THREAD_THRESHOLD_PCT))) {
+      tsdb.getConfig().register(getConfigKey(THREAD_THRESHOLD_PCT), 85, false,
+          "Percentage of active threads to consider before applying the reschedule. "
+              + "Can be null to use the default.");
+    }
+
+    if (!tsdb.getConfig().hasProperty(getConfigKey(MAX_THREAD_PER_USER_PCT))) {
+      tsdb.getConfig().register(getConfigKey(MAX_THREAD_PER_USER_PCT), 50, false,
+          "Percentage of active threads per user to consider before applying the reschedule when the total "
+              + "percentage of active thread in the executor falls below the "
+              + THREAD_THRESHOLD_PCT + " threshold. " + "Can be null to use the default.");
+    }
+
+    int threadThresholdPct = tsdb.getConfig().getInt(getConfigKey(THREAD_THRESHOLD_PCT));
+    if (threadThresholdPct > 100) {
+      return Deferred.fromError(
+          new IllegalStateException(THREAD_THRESHOLD_PCT + " value should be less than 100"));
+    }
+    int threadThresholdPerUserPct = tsdb.getConfig().getInt(getConfigKey(THREAD_THRESHOLD_PCT));
+    if (threadThresholdPerUserPct > 100) {
+      return Deferred.fromError(
+          new IllegalStateException(MAX_THREAD_PER_USER_PCT + " value should be less than 100"));
+    }
+
+    Integer maxSize = tsdb.getConfig().getInt(getConfigKey(QUEUE_MAX_SIZE));
+    corePoolSize = tsdb.getConfig().getInt(getConfigKey(CORE_THREAD_POOL_SIZE));
+    maxTPoolSize = tsdb.getConfig().getInt(getConfigKey(MAX_THREAD_POOL_SIZE));
+    threadThresholdCnt = maxTPoolSize * (threadThresholdPct / 100);
+    threadThresholdPerUserCnt = maxTPoolSize * (threadThresholdPerUserPct / 100);
+
+    LOG.info(
+        "Initializing new UserAwareThreadPoolExecutor with max queue size of {} with core threads {} and a "
+            + "maximum pool size of {} with threshold Thread count of {}.",
+        maxSize, corePoolSize, maxTPoolSize, threadThresholdPct);
+
+    workQueue = new LinkedBlockingQueue<Runnable>(maxSize);
+
+    // TODO: better alternative?
+    // If the worker queue is full, we run the task in the thread that calls execute(task)
+    RejectedExecutionHandler inPlaceRunTaskHandler = new RejectedExecutionHandler() {
+
+      @Override
+      public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+        if (!executor.isShutdown()) {
+          r.run();
```

Comment:
Yeah thats ok I think but log if we skip the executor.
And throw an error if the executor is indeed shut down as that means we closed the executor too early.

Reply:
Also, to answer you other question about what happens when the queue is full, the task gets executed in this handler and will run in the same thread that tries to run but rejects it since the queue is full. I'll log it when this happens.